src/modelfree/transparent.py

```diff
@@ -0,0 +1,188 @@
+from abc import ABC, abstractmethod
+
+import gym
+from gym_compete.policy import LSTMPolicy, MlpPolicyValue
+import numpy as np
+from stable_baselines.common.policies import FeedForwardPolicy, nature_cnn
+import tensorflow as tf
+
+from aprl.envs.multi_agent import CurryVecEnv, _tuple_pop, _tuple_space_augment
+
+TRANSPARENCY_KEYS = ('obs', 'ff_policy', 'ff_value', 'hid')
+
+
+class TransparentPolicy(ABC):
+    """Policy which returns its observations and/or activations in its call to self.predict
+
+    :param transparent_params: dict with potential keys in TRANSPARENCY_KEYS
+    - If key is not present, then we don't provide this data as part of the data dict in step.
+    - If key is present, value (bool) corresponds to whether we augment the observation space
+    with it. This is because TransparentCurryVecEnv needs this information to modify its
+    observation space, and we want all of the transparency-related parameters in one dict.
+    """
+    def __init__(self, transparent_params):
+        if transparent_params is None:
+            raise ValueError("TransparentPolicy requires transparent_params.")
+        for key in transparent_params:
+            if key not in TRANSPARENCY_KEYS:
+                raise KeyError(f"Unrecognized transparency key: {key}")
+        self.transparent_params = transparent_params
+
+    @abstractmethod
+    def get_obs_sizes(self):
+        """List of dimensionalities of different data that could be appended to an observation.
+        Specifically (observations, (concatenated) policy acts., value acts., and hidden state)
+
+        This is used in get_obs_aug_amount, which has the common logic for all subclasses.
+        If class does not support a key, its value should be None.
+
+        :return: (list<int,None>) sizes of data from TRANSPARENCY_KEYS for this policy.
+        """
+        raise NotImplementedError()
+
+    def get_obs_aug_amount(self):
+        """Execute loop for calculating amount by which to augment observations"""
+        obs_aug_amount = 0
+        obs_sizes = self.get_obs_sizes()
+        for key, val in list(zip(TRANSPARENCY_KEYS, obs_sizes)):
+            if self.transparent_params.get(key):
+                obs_aug_amount += val
+        return obs_aug_amount
+
+    def get_transparency_dict(self, transparent_objs):
+        """Construct dictionary of transparent data
+
+        :param transparent_objs: (tuple<np.ndarray,None>) values for keys in TRANSPARENCY_KEYS
+        :return: (dict[str, np.ndarray]) dict with all exposed data
+        """
+        transparency_dict = {k: np.squeeze(v) for k, v in zip(TRANSPARENCY_KEYS, transparent_objs)
+                             if k in self.transparent_params}
```

Comment:
Agree with something similar to an Adam comment above; I feel like this would be cleaner if you just iterated over the keys present in self.transparent_params, rather than iterating over the full list of TRANSPARENCY_KEYS with a conditional subset condition